### DISCUSSION TITLE
PerspectiveWidget gets in unrecoverable state

### DISCUSSION THREAD

**Discussion by robambalu** [perspective-dev/perspective#1400] [Q&A] [answered]
Reproduce the issue:
- create a multi-line d3_y_line graph
- go to the edit settings, change view to datagrid
-  go back to y-line 

at this point the graph will only show 1 of the lines, and going into edit wont let you choose multi lines again either ( seems to be in an unrecoverable state at this point )

sample graph:
import pandas
x = pandas.DataFrame( { "a" : list(range(10)), "b" : list(x*x for x in range(10))})
perspective.PerspectiveWidget( x, plugin='d3_y_line')

after going through3 steps above:
![image](https://user-images.githubusercontent.com/28680700/116099710-3c21c880-a67a-11eb-882c-104bb3cd608b.png)

**Comment by texodus** [comment, marked as answer]
Use `shift+click` on the inactive columns to additional columns when in chart mode, or drag/drop them from bottom to top (this is just the reverse of the behavior of `click` and `shift+click` in Datagrid mode).

![bug](https://user-images.githubusercontent.com/60666/116184966-b68c3000-a6ee-11eb-8622-66e83ae75738.gif)


There is a bug here in `PerspectiveWidget` though - setting `plugin` to `"d3_y_line"` without also explicitly setting `columns` should _not_ default to all columns (as it does currently), but rather should display the behavior shown in your screenshot, e.g. choose a single numeric column if one is not already in the 1st position of the column selector.  To get the chart you want, the (intended) API would be: 

```python
import pandas
x = pandas.DataFrame( { "a" : list(range(10)), "b" : list(x*x for x in range(10))})
perspective.PerspectiveWidget( x, plugin='d3_y_line', columns=["a", "b", "index"])
```

**Comment by robambalu** [comment]
I see, ctrl+shift does seem to let me add them back in, but its odd to see it drop them all by switching to datagrid and then back ( also, I may be missing it, but why the need for ctrl+?)

**Comment by texodus** [comment]
I suspect this would be more obvious on a non-trivial example with more than 3 numeric columns, e.g. [Superstore](https://bl.ocks.org/texodus/803de90736a3641ad91c5c7a1b49d0a7) example has 21 columns, only 5 of which are numeric and only 2 of those share a logic axis.  Perspective will not change arguments supplied explicitly as `kwargs` but fills in other parameters with sane defaults to prevent overdraw, and will modify parameters using these defaults when the context changes due to user interaction.

`Shift+` (not `Ctrl+`) just toggles the default behavior - in `Chart` mode it makes the UI act like a `DataGrid`, and vice-versa.  Drag/drop works the same in both modes.

**Reply by robambalu** [reply]
understood, thanks Andrew
